src/placement.c

```diff
@@ -0,0 +1,548 @@
+{
+	assert(bmp);
+	assert(view);
+
+	struct server *server = view->server;
+	assert(server);
+
+	/* Always start with a fresh bitmap */
+	destroy_bitmap(bmp);
+
+	struct output *output = view->output;
+	if (!output_is_usable(output)) {
+		return;
+	}
+
+	int nviews = count_views(view);
+	if (nviews < 1) {
+		return;
+	}
+
+	/* Number of rows/columns is bounded by two per view plus screen edges */
+	int max_rc = 2 * nviews + 2;
+
+	int rows[max_rc];
+	int cols[max_rc];
+
+	/* First edges of grid are start of usable area of output */
+	struct wlr_box usable = output_usable_area_in_layout_coords(output);
+	int usable_right = usable.x + usable.width;
+	int usable_bottom = usable.y + usable.height;
+
+	cols[0] = usable.x;
+	rows[0] = usable.y;
+
+	cols[1] = usable_right;
+	rows[1] = usable_bottom;
+
+	int nr_rows = 2;
+	int nr_cols = 2;
+
+	struct view *v;
+	for_each_view(v, &server->views, LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		if (v == view || v->output != output) {
+			continue;
+		}
+
+		struct border margin = ssd_get_margin(v->ssd);
+		int x = v->pending.x - margin.left;
+		int y = v->pending.y - margin.top;
+
+		/* Add a column if the left view edge is in the usable region */
+		if (x > usable.x && x < usable_right) {
+			assert(nr_cols < max_rc);
+			cols[nr_cols++] = x;
+		}
+
+		/* Add a row if the top view edge is in the usable region */
+		if (y > usable.y && y < usable_bottom) {
+			assert(nr_rows < max_rc);
+			rows[nr_rows++] = y;
+		}
+
+		x = v->pending.x + v->pending.width + margin.right;
+		y = v->pending.y + v->pending.height + margin.bottom;
+
+		/* Add a column if the right view edge is in the usable region */
+		if (x > usable.x && x < usable_right) {
+			assert(nr_cols < max_rc);
+			cols[nr_cols++] = x;
+		}
+
+		/* Add a row if the bottom view edge is in the usable region */
+		if (y > usable.y && y < usable_bottom) {
+			assert(nr_rows < max_rc);
+			rows[nr_rows++] = y;
+		}
+	}
+
+	nr_rows = order_grid(rows, nr_rows);
+	nr_cols = order_grid(cols, nr_cols);
+
+	int nr_grid = (nr_rows - 1) * (nr_cols - 1);
+
+	bmp->rows = malloc((nr_rows + nr_cols + nr_grid) * sizeof(int));
+	if (!bmp->rows) {
+		return;
+	}
+
+	bmp->nr_rows = nr_rows;
+	bmp->nr_cols = nr_cols;
+
+	bmp->cols = bmp->rows + nr_rows;
+	bmp->grid = bmp->cols + nr_cols;
+
+	memcpy(bmp->rows, rows, nr_rows * sizeof(int));
+	memcpy(bmp->cols, cols, nr_cols * sizeof(int));
+	memset(bmp->grid, 0, nr_grid * sizeof(int));
+}
+
+/*
+ * Perform a rightmost binary search along a list of edges in a 1-D grid for
+ * the maximum index j such that edges[j] <= val. The list of edges must be
+ * sorted in increasing order.
+ *
+ * For a returned index j:
+ *
+ * - The index j == -1 implies that val < edges[0].
+ * - An index 0 <= j < (nedges - 1) implies that edges[j] <= val < edges[j + 1].
+ * - The index j == (nedges - 1) implies that edges[nedges - 1] <= val.
+ */
+static int
+find_interval(int *edges, int nedges, double val)
```

Comment:
Could we use `bsearch()` for this?
I don't mind keeping it as it is though.

Reply:
I looked into `bsearch`, but needed a custom implementation because `bsearch` returns `NULL` when there is no exactly matching element. The implementation here relies on inexact interval matches to simplify identification of the start and end of the range of intervals that bound a view.

Reply:
Good point!